plugins/tpm/fu-tpm-device.c

```diff
@@ -0,0 +1,266 @@
+/*
+ * Copyright (C) 2019 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <tss2/tss2_esys.h>
+
+#include "fu-tpm-device.h"
+
+struct _FuTpmDevice {
+	FuUdevDevice		 parent_instance;
+};
+
+G_DEFINE_TYPE (FuTpmDevice, fu_tpm_device, FU_TYPE_UDEV_DEVICE)
+
+static void Esys_Finalize_autoptr_cleanup (ESYS_CONTEXT *esys_context)
+{
+	Esys_Finalize (&esys_context);
+}
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (ESYS_CONTEXT, Esys_Finalize_autoptr_cleanup)
+
+static gboolean
+fu_tpm_device_probe (FuUdevDevice *device, GError **error)
+{
+	FuTpmDevice *self = FU_TPM_DEVICE (device);
+
+	/* set the physical ID */
+	if (!fu_udev_device_set_physical_id (device, "tpm", error))
+		return FALSE;
+
+	/* already open, so don't touch */
+	g_object_set (self, "device-file", NULL, NULL);
+	return TRUE;
+}
+
+static gboolean
+fu_plugin_tpm_get_uint32 (ESYS_CONTEXT *ctx, guint32 query, guint32 *val, GError **error)
+{
+	TSS2_RC rc;
+	g_autofree TPMS_CAPABILITY_DATA *capability = NULL;
+
+	g_return_val_if_fail (val != NULL, FALSE);
+
+	rc = Esys_GetCapability (ctx, ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE,
+				 TPM2_CAP_TPM_PROPERTIES, query, 1, NULL, &capability);
+	if (rc != TSS2_RC_SUCCESS) {
+		g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+			     "capability request failed for query %x", query);
+		return FALSE;
+	}
+	if (capability->data.tpmProperties.count == 0) {
+		g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+			     "no properties returned for query %x", query);
+		return FALSE;
+	}
+	if (capability->data.tpmProperties.tpmProperty[0].property != query) {
+		g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+			     "wrong query returned (got %x expected %x)",
+			     capability->data.tpmProperties.tpmProperty[0].property,
+			     query);
+		return FALSE;
+	}
+
+	*val = capability->data.tpmProperties.tpmProperty[0].value;
+	return TRUE;
+}
+
+static gchar *
+fu_plugin_tpm_get_string (ESYS_CONTEXT *ctx, guint32 query, GError **error)
+{
+	guint32 val_be = 0;
+	guint32 val;
+	gchar result[5] = {'\0'};
+
+	/* return four bytes */
+	if (!fu_plugin_tpm_get_uint32 (ctx, query, &val_be, error))
+		return NULL;
+	val = GUINT32_FROM_BE(val_be);
+	memcpy (result, (gchar *) &val, 4);
+
+	/* convert non-ASCII into spaces */
+	for (guint i = 0; i < 4; i++) {
+		if (!g_ascii_isgraph (result[i]) && result[i] != '\0')
+			result[i] = 0x20;
+	}
+
+	return fu_common_strstrip (result);
+}
+
+static const gchar *
+fu_tpm_device_convert_manufacturer (const gchar *manufacturer)
+{
+	if (g_strcmp0 (manufacturer, "AMD") == 0)
```

Comment:
I think it would be useful to reference the spec this was last updated from in a comment so that when a new vendor enters it's more obvious why they're missing.

Reply:
Agree. I've added a link to the spec version.